---
title: "Kinome VCF data analysis"
author: "JW Yi"
date: "Tuesday, February 04, 2015"
output:
  pdf_document:
    highlight: pygments
---

#1. Data import and exploration

```{r, message=FALSE}
library(VariantAnnotation); library(ggplot2)
setwd("D://Research/014 Kinome, PTMC/Work/")
vcf <- readVcf("gatk.snp.multi.thread.vcf", "hg19")
```

## Header Information and Genomic Position   

```{r}
vcf
header(vcf)
head(rowData(vcf), 6)

GT <- geno(vcf)$GT
AD <- geno(vcf)$AD
DP <- geno(vcf)$DP
GQ <- geno(vcf)$GQ
PL <- geno(vcf)$PL

GT[1:6,1:3]
AD[1:6,1:3]
DP[1:6,1:3]
GQ[1:6,1:3]
PL[1:6,1:3]
```

## Distribution of DP(Read Depth)   

```{r}
fivenum(DP) # minimum, first quartile, median, upper quartile, maximum
hist(DP, breaks = seq(0, 200, by=1), main = "DP non-zero values", ylab="DP")
```

## Info data

```{r}
info(vcf)[1:6, ]
```

## Compare quality measures between novel(not in dbSNP) and known(in dbSNP) variants.

```{r, message=FALSE}
library(SNPlocs.Hsapiens.dbSNP.20101109)
rd <- rowData(vcf)
seqlevels(rd)
df <- snpSummary(vcf)
head(df) 
# g00 : Counts for homozygous reference, g01 : heterozygous, g11 : homozygous alternate
# a0Freq : Frequency of reference allele, a1Freq : Frequency of alternate allle
# HWEzscore : Z score from H-W Equilibrium, HWEpvalue : p value from H-W Equilibrium

snps <- getSNPlocs(c("ch1", "ch2", "ch3", "ch4", "ch5", "ch6", "ch7", "ch8", "ch9", "ch10",
                     "ch11", "ch12", "ch13", "ch14", "ch15", "ch16", "ch17", "ch18", "ch19", "ch20",
                     "ch21", "ch22", "chX", "chY", "chMT"), as.GRanges=TRUE)
dbsnp <- sub("rs","", names(rd)) %in% snps$RefSNP_id
table(dbsnp)

metrics <- data.frame(QUAL=qual(vcf), inDbSNP = dbsnp, DP=info(vcf)$DP, MQ = info(vcf)$MQ)
ggplot(metrics, aes(x=MQ, fill=inDbSNP)) +
  geom_density(alpha=0.5) + theme(legend.position="top")
```

#2. Locating Variants in and around Genes

```{r, message=FALSE, warning=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
rd <- rowData(vcf)
loc <- locateVariants(rd, txdb, CodingVariants())
head(loc)
allvar <- locateVariants(rd, txdb, AllVariants())
head(allvar)
```

## Did any coding variants match more than one gene?

```{r}
splt <- split(mcols(loc)$GENEID, mcols(loc)$QUERYID)
table(sapply(splt, function(x) length(unique(x)) > 1))
```

## Summarize the number of coding variants by gene ID

```{r}
splt <- split(mcols(loc)$QUERYID, mcols(loc)$GENEID)
head(sapply(splt, function(x) length(unique(x))))
```

#3. Summarize Variants

## Txdb (Transcripts DB) as query

```{r}
coding1 <- summarizeVariants(txdb, vcf, CodingVariants())
head(colSums(assays(coding1)$counts))
count.coding1 <- as.data.frame(assays(coding1)$counts)
count.coding1[1:10,1:3]
```

## GRangesList as the query

```{r}
cdsbytx <- cdsBy(txdb, "tx")
coding2 <- summarizeVariants(cdsbytx, vcf, CodingVariants())
stopifnot(identical(assays(coding1)$counts, assays(coding2)$counts))  
# Results from two query are identical
```

## Promotor region variants summarized by transcript

```{r}
tx <- transcripts(txdb)
txlst <- splitAsList(tx, seq_len(length(tx)))
promoter <- summarizeVariants(txlst, vcf, PromoterVariants())
head(colSums(assays(promoter)$counts))
```

## Summarize all variants

```{r}
allvariants <- summarizeVariants(txlst, vcf, findOverlaps)
head(colSums(assays(allvariants)$counts))
```

## Write all variants counts into csv file
```{r}
coding1 <- summarizeVariants(txdb, vcf, CodingVariants())
coding2 <- summarizeVariants(txdb, vcf, IntronVariants())
coding3 <- summarizeVariants(txdb, vcf, FiveUTRVariants())
coding4 <- summarizeVariants(txdb, vcf, ThreeUTRVariants())
coding5 <- summarizeVariants(txdb, vcf, SpliceSiteVariants())
coding6 <- summarizeVariants(txdb, vcf, PromoterVariants())

a <- colSums(assays(coding1)$counts)
b <- colSums(assays(coding2)$counts)
c <- colSums(assays(coding3)$counts)
d <- colSums(assays(coding4)$counts)
e <- colSums(assays(coding5)$counts)
f <- colSums(assays(coding6)$counts)
g <- colSums(assays(allvariants)$counts)

variantscounts <- cbind(a,b,c,d,e,f,g)
colnames(variantscounts) <- c("Coding", "Intron", "FiveUTR", "ThreeUTR", "SpliceSite", "Promotor", "All")
head(variantscounts)
write.csv(variantscounts, "variantscounts.csv")
```

#4. Amino Acid Coding Changes

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
coding <- predictCoding(vcf, txdb, seqSource=Hsapiens)
coding
coding [mcols(coding)$CONSEQUENCE=="synonymous"]
coding [mcols(coding)$CONSEQUENCE=="nonsynonymous"]
coding [mcols(coding)$CONSEQUENCE=="nonsense"]

elementMetadata <- as.data.frame(coding@elementMetadata)
seqnames <- as.data.frame(coding@seqnames)
ranges <- as.data.frame(coding@ranges)
strands <- as.data.frame(coding@strand)
write.csv(cbind(seqnames, ranges, strands, elementMetadata), "predictcoding.csv")
```

#5. SIFT(Sorting Intolerant From Tolerant) and PolyPhen(Polymorphism Phenotyping) Database

## Identify the non-synonymous variants and obtain the "rsids"

```{r}
nms <- names(coding)
idx <- mcols(coding)$CONSEQUENCE == "nonsynonymous"
nonsyn <- coding[idx]
names(nonsyn) <- nms[idx]
rsids <- unique(names(nonsyn)[grep("rs", names(nonsyn), fixed=T)])
head(rsids)
length(rsids)
```

## Query the PolyPhen database for pridiction of harmful effect by SNPs.

```{r, message=FALSE}
library(PolyPhen.Hsapiens.dbSNP131)
pp <- select(PolyPhen.Hsapiens.dbSNP131, keys=rsids, 
             cols=c("TRAININGSET", "PREDICTION", "PPH2PROB"))
head(pp[!is.na(pp$PREDICTION), ])
write.csv(pp[!is.na(pp$PREDICTION), ], "prediction.csv")
```

#6. Create a SNP Matrix

## Create a SNP matrix contains allele information

```{r, warning=FALSE}
res <- genotypeToSnpMatrix(vcf)
res

allele2 <- res$map[["allele.2"]]
unique(elementLengths(allele2))  # Number of alternate alleles per variant

res.genotype <- t(as(res$genotype, "character"))
res.genotype[1:6, 1:6]
write.csv(res.genotype, "genotypetosnpmatrix.csv")
```
